libcnb/src/layer/struct_api/mod.rs

```diff
@@ -0,0 +1,235 @@
+
+// Allows to use the layer actions directly.
+impl<T, E> IntoAction<T, (), E> for T {
+    fn into_action(self) -> Result<(T, ()), E> {
+        Ok((self, ()))
+    }
+}
+
+//  Allows to use the layer actions directly wrapped in a Result.
+impl<T, E> IntoAction<T, (), E> for Result<T, E> {
+    fn into_action(self) -> Result<(T, ()), E> {
+        self.map(|value| (value, ()))
+    }
+}
+
+// Allows to use the layer actions directly with a cause as a tuple.
+impl<T, C, E> IntoAction<T, C, E> for (T, C) {
+    fn into_action(self) -> Result<(T, C), E> {
+        Ok(self)
+    }
+}
+
+// Allows to use the layer actions directly with a cause as a tuple wrapped in a Result.
+impl<T, C, E> IntoAction<T, C, E> for Result<(T, C), E> {
+    fn into_action(self) -> Result<(T, C), E> {
+        self
+    }
+}
+
+/// A reference to an existing layer on disk.
+///
+/// Provides functions to modify the layer such as replacing its metadata, environment, SBOMs or
+/// exec.d programs.
+///
+/// To obtain a such a reference, use [`BuildContext::cached_layer`] or [`BuildContext::uncached_layer`].
+pub struct LayerRef<B, MAC, RAC>
+where
+    B: Buildpack + ?Sized,
+{
+    name: LayerName,
+    // Technically not part of the layer itself. However, the functions that modify the layer
+    // will need a reference to the layers directory as they will also modify files outside the
+    // actual layer directory. To make LayerRef nice to use, we bite the bullet and include
+    // the layers_dir here.
+    layers_dir: PathBuf,
+    buildpack: PhantomData<B>,
+    pub state: LayerState<MAC, RAC>,
+}
+
+impl<B, MAC, RAC> LayerRef<B, MAC, RAC>
+where
+    B: Buildpack,
+{
+    /// Returns the path to the layer on disk.
+    pub fn path(&self) -> PathBuf {
+        self.layers_dir.join(self.name.as_str())
+    }
+
+    /// Writes the given layer metadata to disk.
+    ///
+    /// Any existing layer metadata will be overwritten. The new value does not have to be of the
+    /// same type as the existing metadata.
+    pub fn write_metadata<M>(&self, metadata: M) -> crate::Result<(), B::Error>
+    where
+        M: Serialize,
+    {
+        crate::layer::shared::replace_layer_metadata(&self.layers_dir, &self.name, metadata)
+            .map_err(|error| {
+                crate::Error::LayerError(LayerError::WriteLayerError(
+                    WriteLayerError::WriteLayerMetadataError(error),
+                ))
+            })
+    }
+
+    /// Writes the given layer environment to disk.
+    ///
+    /// Any existing layer environment will be overwritten.
+    pub fn write_env(&self, env: impl Borrow<LayerEnv>) -> crate::Result<(), B::Error> {
+        env.borrow()
+            .write_to_layer_dir(self.path())
+            .map_err(|error| {
+                crate::Error::LayerError(LayerError::WriteLayerError(WriteLayerError::IoError(
+                    error,
+                )))
+            })
+    }
+
+    /// Reads the current layer environment from disk.
```

Comment:
Maybe add a note saying this will also add the additional magic entries?

Reply:
Good call!